### PR TITLE
FOGL-1144 - http_north plugin _send() method refactored

### PR DESCRIPTION
**Pull request by amarendra-dianomic**
Test result:
```
tests/unit/python/foglamp/plugins/north/ocs/test_ocs.py ......FF..FFF                                                                     [ 30%]
tests/unit/python/foglamp/plugins/north/omf/test_omf.py .....F..FFFFFFFFFFFFFFFF.........F......                                          [ 33%]
======================================= 23 failed, 1201 passed, 33 skipped, 11 warnings in 97.72 seconds ========================================

```

### CODE REVIEW

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 
 http_north = None
 config = ""
+MAX_ATTEMPTS = 5
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: it should be configurable

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 2]
```diff
@@ -139,33 +129,65 @@ def send_payloads(self, payloads, stream_id):
         """ send a list of block payloads """
         num_count = 0
         last_id = None
-        async with aiohttp.ClientSession() as session:
-            payload_to_be_send = list()
-            for payload in payloads:
-                num_count += 1
-                last_id = payload['id']
+
+        for payload in payloads:
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: I think there is only 1 payload block of readings

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 3]
```diff
@@ -139,33 +129,65 @@ def send_payloads(self, payloads, stream_id):
...
+        else:  # any other value means retry N times
+            attempt_count = MAX_ATTEMPTS + 1
+
+        while attempt_count != 0:
+            try:
+                if raised_exc:
+                    _LOGGER.error('caught "%s" url:%s, remaining tries %s, sleeping %.2fsecs',
+                                  raised_exc, url, attempt_count, backoff_interval)
+                    import time
+                    time.sleep(backoff_interval)
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: it should be async sleep

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 4]
```diff
@@ -139,33 +129,65 @@ def send_payloads(self, payloads, stream_id):
...
+                raised_exc = FailedRequest(message=exc, url=url, raised=exc.__class__.__name__)
+            else:
+                raised_exc = None
+                break
+
+            attempt_count -= 1
+
+        if raised_exc:
+            raise raised_exc
+        else:
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: we can avoid else

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 5]
```diff
@@ -80,56 +81,45 @@ def plugin_send(data, payload, stream_id):
...
-
-        After waiting for a threshold limit i.e. config["wait_to_shutdown"]
-
-        done = sent successfully + got an error + trapped into exception
-        pending =  the requests (tasks) waiting for response Or in queue
-
-        """
-        self.event_loop.run_until_complete(self.cancel_tasks())
-
-    async def cancel_tasks(self):
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: what if a task is terminated or asked for graceful cancelled

File: python/foglamp/plugins/north/http_north/http_north.py (hunk 4 above)
Comment (amarendra-dianomic) [reply python/foglamp/plugins/north/http_north/http_north.py]: Just for self documentation purpose.

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 6]
```diff
@@ -139,33 +137,63 @@ def send_payloads(self, payloads, stream_id):
         """ send a list of block payloads """
         num_count = 0
         last_id = None
-        async with aiohttp.ClientSession() as session:
-            payload_to_be_send = list()
-            for payload in payloads:
-                num_count += 1
-                last_id = payload['id']
+        for payload in payloads:
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: why we are looping through? and are not sending a full block?!

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 7]
```diff
@@ -139,33 +137,63 @@ def send_payloads(self, payloads, stream_id):
...
+            except (aiohttp.ClientError, web.HTTPClientError, web.HTTPServerError) as exc:
+                raised_exc = FailedRequest(message=exc, url=url, raised=exc.__class__.__name__)
+            else:
+                raised_exc = None
+                break
+
+            attempt_count -= 1
+
+        if raised_exc:
+            raise raised_exc
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: in this case, have we checked the stats (not sent) are being handled properly?

File: python/foglamp/plugins/north/http_north/http_north.py [hunk 8]
```diff
@@ -139,33 +137,63 @@ def send_payloads(self, payloads, stream_id):
...
+            attempt_count = 1
+        else:  # any other value means retry N times
+            attempt_count = max_attempts + 1
+
+        while attempt_count != 0:
+            try:
+                if raised_exc:
+                    _LOGGER.error('caught "%s" url:%s, remaining tries %s, sleeping %.2fsecs',
+                                  raised_exc, url, attempt_count, backoff_interval)
+                    await asyncio.sleep(backoff_interval)
```
Comment (praveen-garg) [inline python/foglamp/plugins/north/http_north/http_north.py]: `await asyncio.sleep(backoff_interval, loop=self.event_loop)`

**Comment by ashish-jabble** [comment]
@amarendra-dianomic As Fogl-1366 changes are in, please resolve conflicts and fix unit tests as well.

**Comment by amarendra-dianomic** [comment]
No longer needed as it will eventually be replaced by [FOGL-1425](https://scaledb.atlassian.net/browse/FOGL-1425).

**Pull request closed without merging**